### ISSUE TITLE
Use same semantics for line_to as the Canvas

### ISSUE THREAD

**Issue by HalfVoxel** [nical/lyon#555]
```
The lineTo(x, y) method, when invoked, must run these steps:

    If either of the arguments are infinite or NaN, then return.

    If the object's path has no subpaths, then ensure there is a subpath for (x, y).

    Otherwise, connect the last point in the subpath to the given point (x, y) using a straight line, and then add the given point (x, y) to the subpath.
```
From https://html.spec.whatwg.org/multipage/canvas.html#dom-context-2d-lineto

The important part is that lineTo is identical to moveTo if the current path is empty. This would do away with the (imo) kinda awkward move_to_if_needed method which makes a line_to for an empty path draw a line from (0,0) to that point (I'm not sure when this is ever what you want).

**Comment by nical** [comment]
The [SVG spec](https://www.w3.org/TR/SVG/paths.html#PathDataClosePathCommand) says:

> If a "closepath" is followed immediately by a "moveto", then the "moveto" identifies the start point of the next subpath. If a "closepath" is followed immediately by any other command, then the next subpath starts at the same initial point as the current subpath.

Which I think is reasonable. I agree, though, that when the path is empty, defaulting to zero isn't very good and turning the line_to into a move_to would be better in that case.

**Comment by nical** [comment]
It's unfortunate that SVG and Canvas2D handle empty sub paths differently.

I am musing with the idea of separating the path building and sanitization. For example having a `PathBuilder` that only has debug assertions checking for NaNs and that sub paths are appropriately started, and optionally wrap them into sanitizers that insert the proper commands when needed. This way the sanitization code would not have to be repeated between the different path data structures and we could have differen path sanitizers (for example one that follows the SVG spec and one that follows Canvas2D).

For example:

```rust
pub type CanvasPathBuilder = CanvasSanitizer<lyon::path::Builder>;`
pub type SvgPathBuilder = SvgSanitizer<lyon::path::Builder>;`
```

I want to be able to return endpoint ids when building the segments. Doing nothing when a point contains a `NaN` means `line_to` would need to return `Option<EndpointId>` which is annoying when you have guaranteed beforehand that there is no `NaN`. A canvas-style sanitizer, though, could return `Option<EndpointId>` without enforcing that on other sanitization styles (though they wouldn't be able to share a common interface).

**Comment by HalfVoxel** [comment]
> Doing nothing when a point contains a NaN means line_to would need to return Option<EndpointId> which is annoying when you have guaranteed beforehand that there is no NaN.
Yeah I think silently doing nothing for NaNs or returning an Option is a bad approach too. Javascript kinda has to just ignore it because it's javascript. But I think the right approach is to panic if you try to draw a NaN.

Hmmm... I'm not sure about sanitizers. I think it's better to have a well defined API which is nice to use. Then if you for example write an importer for svg you could handle any api differences there.
Sanitizers feel like they would just increase the API complexity a lot for very little gain.

**Comment by nical** [comment]
> I think that seems reasonable too. And when reading the Canvas spec carefully (and then actually testing it) I notice they do the same thing. Note that it only converts a line_to to a move_to iff the object's path has no subpaths. Note subpaths plural. So if there is an existing subpath (even if it is closed) then it will do a line_to from the last point.

That is reassuring. 

SVG appears to consider the whole path invalid if it does not start with a moveto command.

> I think it's better to have a well defined API which is nice to use. Then if you for example write an importer for svg you could handle any api differences there.
Sanitizers feel like they would just increase the API complexity a lot for very little gain.

Yeah. Since the difference between canvas and SVG appear to be only for the very first command of a path without moveto, having multiple sanitizng behaviors is overkill. Anyone can wrap PathBuilder and do their own sanitizer to match their needs, and it turns out I can't implement canvas's behavior exactly.

The issue with canvas is that it defaults to moving to the first control point of the curve in the case of a bézier segment.
Lyon makes an important distinction between endpoints and control points. They have different ID types and more importantly the endpoints can have custom attributes but control points cannot.
When building a path with custom attributes it's not possible to default to moving to a control point the way canvas would do for quadratic and bézier segments, while it would be possible to default to moving to the end of the segment (works because it is an endpoint and not a control point).

It's too bad that canvas wasn't specified that way. In my opinion it doesn't make sense to have the path pass through a control point while it would be sensible to replace the segment with a moveto to its end, which is equivalent to removing the segment.

Long story short lyon can't do exactly what canvas does (when an empty path starts with a bézier segment).

In lyon we can either:

 * panic
 * default to zero (which we agree isn't very good).
 * make the whole path builder invalid and build an empty path. That's what SVG does but I am not ver fond of it, might as well panic.
 * replace the segment with a `move_to(end_of_segment)`. This is my preferred solution, it's similar to canvas but not exactly the same.

I'm probably going to do the latter. I think it is what makes most sense under lyon's constraints (and in general as well).

**Comment by nical** [comment]
I experimented with that in a branch https://github.com/nical/lyon/commit/7d3c75479fd3125d55b7340a7fcb9b0e89e62fed

**Comment by HalfVoxel** [comment]
> The issue with canvas is that it defaults to moving to the first control point of the curve in the case of a bézier segment.

Yeah that makes no sense for curve_to/quadratic_bezier_to/etc.. In my opinion that seems more like duct tape in order to avoid having it throw an exception in any circumstance. I think it's much better to have it panic in that case.
I wouldn't be against having it panic if one calls line_to without an existing path either. It makes for a slightly nicer to use API to not have it move_to instead, but having it panic feels a bit more *correct*.

> replace the segment with a move_to(end_of_segment). This is my preferred solution, it's similar to canvas but not exactly the same.
Do you mean that if one does a quadratic_bezier_to(ctrl1, ctrl2, to) when the path is empty you just do a move_to(to)? That seems odd? I mean obviously the user intended some kind of quadratic curve. This means you just silently ignore it. Imo it would be better to panic.

**Comment by nical** [comment]
> Do you mean that if one does a quadratic_bezier_to(ctrl1, ctrl2, to) when the path is empty you just do a move_to(to)?

Yes. This is equivalent to ignoring the first edge which was missing an initial position but being in the right state to handle following ones, this applies to curves and lines alike.

> I guess one could make the same argument for line_to though...

Indeed, I don't think lines and curves should be treated differently, here.

> Imo it would be better to panic.

That's also possible. It's a bit odd to me that omitting move_to between sub-paths would work while it omitting it at the very beginning of the path would cause a panic, but I think I could live with it. The advantage of doing that, though is that it avoids the overhead of the boilerplate in each command to handle a supposedly rare case. Panicking is a bit harsh, though. Other vector graphics API handle this one way or another without crashing.

For reference, I looked at cairo and skia: the former does the weird canvas 2d bandaid while the latter moves to (0, 0) even for subsequent sub-paths that don't start with a move command.

**Comment by nical** [comment]
On master (0.16, not published yet), the builder interfaces are separated into the low-level `PathBuilder` trait which requires all sub-paths to be started and ended explicitly (debug assertions), and the `SvgPathBuilder` trait implemented on top of it which has all of the path sanitization and discards the first edge of the path if it isn't a moveto command as I described earlier.
